File: zero_waste_frontend/lib/messages.dart
import 'package:moor/moor.dart';
import 'package:provider/single_child_widget.dart';
import 'proto/home.pb.dart';
import 'proto/zero_waste.pb.dart';
import 'package:flutter_persistent_socket/communication/socket_messages.dart';
import 'proto/containers.pb.dart';
import 'package:provider/provider.dart';
import 'proto/rewards.pb.dart';
import 'package:flutter_persistent_socket/communication/socket_api.dart';

class RxHomeInfo extends SocketRxMessage {
  static const String type = 'home-info';
  final HomeInfo data = HomeInfo();
  final Duration cache = const Duration(days: 365, hours: 0, minutes: 0, seconds: 0);

  RxHomeInfo([SocketRxMessageData message]) : super(type, message);

  @override
  RxHomeInfo fromMessage(SocketRxMessageData message) => RxHomeInfo(message);
}


class TxScanWasteContainer extends SocketTxMessage {
  static const String type = 'scan-container';
  final ScanWasteContainer proto;
  
  
  const TxScanWasteContainer([this.proto]) : super(type, authRequired: true);
  
  static ScanWasteContainer get newProto => ScanWasteContainer();
  
  static TxScanWasteContainer create([ScanWasteContainer Function(ScanWasteContainer data) setData]) => TxScanWasteContainer((setData ?? (p) => p)(TxScanWasteContainer.newProto));
}


class TxGetMagicLink extends SocketTxMessage {
  static const String type = 'get-magic-link';
  final GetMagicLink proto;
  
  
  const TxGetMagicLink([this.proto]) : super(type, authRequired: true);
  
  static GetMagicLink get newProto => GetMagicLink();
  
  static TxGetMagicLink create([GetMagicLink Function(GetMagicLink data) setData]) => TxGetMagicLink((setData ?? (p) => p)(TxGetMagicLink.newProto));
}


class RxContainerPurchases extends SocketRxMessage {
  static const String type = 'container-purchases';
  final ContainerPurchases data = ContainerPurchases();
  

  RxContainerPurchases([SocketRxMessageData message]) : super(type, message);

  @override
  RxContainerPurchases fromMessage(SocketRxMessageData message) => RxContainerPurchases(message);
}


class RxUserProfileData extends SocketRxMessage {
  static const String type = 'user-profile-data';
  final UserProfileData data = UserProfileData();
  

  RxUserProfileData([SocketRxMessageData message]) : super(type, message);

  @override
  RxUserProfileData fromMessage(SocketRxMessageData message) => RxUserProfileData(message);
}


class RxRewards extends SocketRxMessage {
  static const String type = 'rewards';
  final Rewards data = Rewards();
  final Duration cache = const Duration(days: 365, hours: 0, minutes: 0, seconds: 0);

  RxRewards([SocketRxMessageData message]) : super(type, message);

  @override
  RxRewards fromMessage(SocketRxMessageData message) => RxRewards(message);
}


class RxContainerInfo extends SocketRxMessage {
  static const String type = 'container-info';
  final ContainerInfo data = ContainerInfo();
  

  RxContainerInfo([SocketRxMessageData message]) : super(type, message);

  @override
  RxContainerInfo fromMessage(SocketRxMessageData message) => RxContainerInfo(message);
}


class TxLoadContainerInfo extends SocketTxMessage {
  static const String type = 'load-container-info';
  final LoadContainerInfo proto;
  
  
  const TxLoadContainerInfo([this.proto]) : super(type, authRequired: true);
  
  static LoadContainerInfo get newProto => LoadContainerInfo();
  
  static TxLoadContainerInfo create([LoadContainerInfo Function(LoadContainerInfo data) setData]) => TxLoadContainerInfo((setData ?? (p) => p)(TxLoadContainerInfo.newProto));
}


class TxLoadRewards extends SocketTxMessage {
  static const String type = 'load-rewards';
  final LoadRewards proto;
  
  
  const TxLoadRewards([this.proto]) : super(type, authRequired: true);
  
  static LoadRewards get newProto => LoadRewards();
  
  static TxLoadRewards create([LoadRewards Function(LoadRewards data) setData]) => TxLoadRewards((setData ?? (p) => p)(TxLoadRewards.newProto));
}


class RxContainerImpact extends SocketRxMessage {
  static const String type = 'container-impact';
  final ContainerImpact data = ContainerImpact();
  

  RxContainerImpact([SocketRxMessageData message]) : super(type, message);

  @override
  RxContainerImpact fromMessage(SocketRxMessageData message) => RxContainerImpact(message);
}


class TxGetHomeInfo extends SocketTxMessage {
  static const String type = 'get-home-info';
  final GetHomeInfo proto;
  
  
  const TxGetHomeInfo([this.proto]) : super(type, authRequired: true);
  
  static GetHomeInfo get newProto => GetHomeInfo();
  
  static TxGetHomeInfo create([GetHomeInfo Function(GetHomeInfo data) setData]) => TxGetHomeInfo((setData ?? (p) => p)(TxGetHomeInfo.newProto));
}


class RxLoginMagicLinkStatus extends SocketRxMessage {
  static const String type = 'login-magic-link-status';
  final LoginMagicLinkStatus data = LoginMagicLinkStatus();
  

  RxLoginMagicLinkStatus([SocketRxMessageData message]) : super(type, message);

  @override
  RxLoginMagicLinkStatus fromMessage(SocketRxMessageData message) => RxLoginMagicLinkStatus(message);
}


class RxContainerMass extends SocketRxMessage {
  static const String type = 'container-mass';
  final ContainerMass data = ContainerMass();
  

  RxContainerMass([SocketRxMessageData message]) : super(type, message);

  @override
  RxContainerMass fromMessage(SocketRxMessageData message) => RxContainerMass(message);
}


class TxLoadContainerMass extends SocketTxMessage {
  static const String type = 'load-container-mass';
  final LoadContainerMass proto;
  
  
  const TxLoadContainerMass([this.proto]) : super(type, authRequired: true);
  
  static LoadContainerMass get newProto => LoadContainerMass();
  
  static TxLoadContainerMass create([LoadContainerMass Function(LoadContainerMass data) setData]) => TxLoadContainerMass((setData ?? (p) => p)(TxLoadContainerMass.newProto));
}


class RxScannedWasteContainer extends SocketRxMessage {
  static const String type = 'scanned-container';
  final ScannedWasteContainer data = ScannedWasteContainer();
  

  RxScannedWasteContainer([SocketRxMessageData message]) : super(type, message);

  @override
  RxScannedWasteContainer fromMessage(SocketRxMessageData message) => RxScannedWasteContainer(message);
}


class TxLoadContainerImpact extends SocketTxMessage {
  static const String type = 'load-container-impact';
  final LoadContainerImpact proto;
  
  
  const TxLoadContainerImpact([this.proto]) : super(type, authRequired: true);
  
  static LoadContainerImpact get newProto => LoadContainerImpact();
  
  static TxLoadContainerImpact create([LoadContainerImpact Function(LoadContainerImpact data) setData]) => TxLoadContainerImpact((setData ?? (p) => p)(TxLoadContainerImpact.newProto));
}


class TxLoadContainerPurchases extends SocketTxMessage {
  static const String type = 'load-container-purchases';
  final LoadContainerPurchases proto;
  
  
  const TxLoadContainerPurchases([this.proto]) : super(type, authRequired: true);
  
  static LoadContainerPurchases get newProto => LoadContainerPurchases();
  
  static TxLoadContainerPurchases create([LoadContainerPurchases Function(LoadContainerPurchases data) setData]) => TxLoadContainerPurchases((setData ?? (p) => p)(TxLoadContainerPurchases.newProto));
}


class RxWasteContainers extends SocketRxMessage {
  static const String type = 'waste-containers';
  final WasteContainers data = WasteContainers();
  

  RxWasteContainers([SocketRxMessageData message]) : super(type, message);

  @override
  RxWasteContainers fromMessage(SocketRxMessageData message) => RxWasteContainers(message);
}


class TxLoginMagicLink extends SocketTxMessage {
  static const String type = 'login-magic-link';
  final LoginMagicLink proto;
  
  
  const TxLoginMagicLink([this.proto]) : super(type, authRequired: true);
  
  static LoginMagicLink get newProto => LoginMagicLink();
  
  static TxLoginMagicLink create([LoginMagicLink Function(LoginMagicLink data) setData]) => TxLoginMagicLink((setData ?? (p) => p)(TxLoginMagicLink.newProto));
}


List<SocketRxMessage> rxMessages = [
  RxHomeInfo(),
  RxContainerPurchases(),
  RxUserProfileData(),
  RxRewards(),
  RxContainerInfo(),
  RxContainerImpact(),
  RxLoginMagicLinkStatus(),
  RxContainerMass(),
  RxScannedWasteContainer(),
  RxWasteContainers()
];
    
List<SingleChildWidget> getMessageProviders(SocketApi api) => [
      StreamProvider<RxHomeInfo>(
        create: (c) => _getMessageHandler(api, RxHomeInfo()),
        lazy: false,
      ),
      StreamProvider<RxContainerPurchases>(
        create: (c) => _getMessageHandler(api, RxContainerPurchases()),
        lazy: false,
      ),
      StreamProvider<RxUserProfileData>(
        create: (c) => _getMessageHandler(api, RxUserProfileData()),
        lazy: false,
      ),
      StreamProvider<RxRewards>(
        create: (c) => _getMessageHandler(api, RxRewards()),
        lazy: false,
      ),
      StreamProvider<RxContainerInfo>(
        create: (c) => _getMessageHandler(api, RxContainerInfo()),
        lazy: false,
      ),
      StreamProvider<RxContainerImpact>(
        create: (c) => _getMessageHandler(api, RxContainerImpact()),
        lazy: false,
      ),
      StreamProvider<RxLoginMagicLinkStatus>(
        create: (c) => _getMessageHandler(api, RxLoginMagicLinkStatus()),
        lazy: false,
      ),
      StreamProvider<RxContainerMass>(
        create: (c) => _getMessageHandler(api, RxContainerMass()),
        lazy: false,
      ),
      StreamProvider<RxScannedWasteContainer>(
        create: (c) => _getMessageHandler(api, RxScannedWasteContainer()),
        lazy: false,
      ),
      StreamProvider<RxWasteContainers>(
        create: (c) => _getMessageHandler(api, RxWasteContainers()),
        lazy: false,
      )
];

Stream<T> _getMessageHandler<T extends SocketRxMessage>(SocketApi api, T message) {
  Stream<T> s = api.getMessageHandler(message);
  api.fireFromCache(message);
  return s;
}

File: zero_waste_backend/proto/messages.py
from proto.form_errors import *
from proto.home import *
from proto.sfiles import *
from proto.socket_api import *
from proto.authentication import *
from proto.uploader import *
from abc import ABC
from proto.zero_waste import *
from proto.containers import *
import dataclasses
from proto.rewards import *
from betterproto import *


class RxMessageData:
    def __init__(self, json_data: dict):
        self.headers = json_data.get('headers')
        self.body = json_data.get('body')
        self.apiVersion = self.headers.get('apiVersion')
        self.authHeader = self.headers.get('authHeader')
        self.ack = self.headers.get('ack')
        self.uuid = self.headers.get('uuid')
        self.type = self.headers.get('messageType')
        self.retryCount = self.headers.get('retryCount', 0)


class RxMessage(ABC):
    proto = None
    type = None
    auth_required = True

    def __init__(self, data: RxMessageData, user=None):
        self.data = data
        self.set_data(data)
        self.user = user

    def set_data(self, data: RxMessageData):
        self.proto = self.proto().from_dict(data.body)

class TxMessage(ABC):
    proto: betterproto.Message = None
    type: str = None

    def __init__(self, proto=None):
        self.fields = {}
        if proto is not None:
            assert type(proto) is self.proto
            self.proto = proto
        else:
            self.proto = self.proto()

    def get_message(self) -> dict:
        return {
            'headers': {
                'messageType': self.type,
            },
            'body': self.proto.to_dict()
        }

def to_dict_patch(
        self, casing: Casing = Casing.CAMEL, include_default_values: bool = False
) -> dict:
    """
    Returns a dict representation of this message instance which can be
    used to serialize to e.g. JSON. Defaults to camel casing for
    compatibility but can be set to other modes.

    `include_default_values` can be set to `True` to include default
    values of fields. E.g. an `int32` type field with `0` value will
    not be in returned dict if `include_default_values` is set to
    `False`.
    """
    output: Dict[str, Any] = {}
    for field in dataclasses.fields(self):
        meta = FieldMetadata.get(field)
        v = getattr(self, field.name)
        cased_name = str(meta.number)
        if meta.proto_type == "message":
            if isinstance(v, datetime):
                if v != DATETIME_ZERO or include_default_values:
                    output[cased_name] = _Timestamp.timestamp_to_json(v)
            elif isinstance(v, timedelta):
                if v != timedelta(0) or include_default_values:
                    output[cased_name] = _Duration.delta_to_json(v)
            elif meta.wraps:
                if v is not None or include_default_values:
                    output[cased_name] = v
            elif isinstance(v, list):
                # Convert each item.
                v = [i.to_dict(casing, include_default_values) for i in v]
                if v or include_default_values:
                    output[cased_name] = v
            else:
                if v._serialized_on_wire or include_default_values:
                    output[cased_name] = v.to_dict(casing, include_default_values)
        elif meta.proto_type == "map":
            for k in v:
                if hasattr(v[k], "to_dict"):
                    v[k] = v[k].to_dict(casing, include_default_values)

            if v or include_default_values:
                output[cased_name] = [{'1': k, '2': v} for k, v in v.items()]
        elif v != self._get_field_default(field, meta) or include_default_values:
            if meta.proto_type in INT_64_TYPES:
                if isinstance(v, list):
                    output[cased_name] = [str(n) for n in v]
                else:
                    output[cased_name] = str(v)
            elif meta.proto_type == TYPE_BYTES:
                if isinstance(v, list):
                    output[cased_name] = [b64encode(b).decode("utf8") for b in v]
                else:
                    output[cased_name] = b64encode(v).decode("utf8")
            elif meta.proto_type == TYPE_ENUM:
                if isinstance(v, list):
                    output[cased_name] = [e for e in v]
                else:
                    output[cased_name] = v
            else:
                output[cased_name] = v
    return output


def from_dict_patch(self: T, value: dict) -> T:
    """
    Parse the key/value pairs in `value` into this message instance. This
    returns the instance itself and is therefore assignable and chainable.
    """
    self._serialized_on_wire = True
    fields_by_name = {str(FieldMetadata.get(f).number): f for f in dataclasses.fields(self)}
    for key in value:
        snake_cased = safe_snake_case(key)
        if snake_cased in fields_by_name:
            field = fields_by_name[snake_cased]
            meta = FieldMetadata.get(field)

            if value[key] is not None:
                if meta.proto_type == "message":
                    v = getattr(self, field.name)
                    if isinstance(v, list):
                        cls = self._betterproto.cls_by_field[field.name]
                        for i in range(len(value[key])):
                            v.append(cls().from_dict(value[key][i]))
                    elif isinstance(v, datetime):
                        v = datetime.fromisoformat(
                            value[key].replace("Z", "+00:00")
                        )
                        setattr(self, field.name, v)
                    elif isinstance(v, timedelta):
                        v = timedelta(seconds=float(value[key][:-1]))
                        setattr(self, field.name, v)
                    elif meta.wraps:
                        setattr(self, field.name, value[key])
                    else:
                        v.from_dict(value[key])
                elif meta.map_types and meta.map_types[1] == TYPE_MESSAGE:
                    v = getattr(self, field.name)
                    cls = self._betterproto.cls_by_field[field.name + ".value"]
                    for k in value[key]:
                        # fixme - map is converted to ['1': key, '2': value], which isn't consistent with betterproto
                        v[k] = cls().from_dict(value[key][k])
                else:
                    v = value[key]
                    if meta.proto_type in INT_64_TYPES:
                        if isinstance(value[key], list):
                            v = [int(n) for n in value[key]]
                        else:
                            v = int(value[key])
                    elif meta.proto_type == TYPE_BYTES:
                        if isinstance(value[key], list):
                            v = [b64decode(n) for n in value[key]]
                        else:
                            v = b64decode(value[key])
                    elif meta.proto_type == TYPE_ENUM:
                        # already in correct format as received
                        pass
                    elif meta.proto_type == TYPE_MAP:
                        v = {o['1']: o['2'] for o in value[key]}


                    if v is not None:
                        setattr(self, field.name, v)
    return self


def init_default_gen_patch(self):
    default_gen = {}

    for field in dataclasses.fields(self.cls):
        meta = FieldMetadata.get(field)
        if meta.proto_type == TYPE_MESSAGE:
            default_gen[field.name] = self.cls._get_field_default_gen(field, meta)
        else:
            default_gen[field.name] = lambda: None

    self.default_gen = default_gen


betterproto.Message.to_dict = to_dict_patch
betterproto.Message.from_dict = from_dict_patch
betterproto.ProtoClassMetadata.init_default_gen = init_default_gen_patch


class TxLoginToken(TxMessage):
    type = 'login-token'
    proto = LoginToken
    auth_required = True
    permissions = [None]


class TxLoginError(TxMessage):
    type = 'login-error'
    proto = LoginError
    auth_required = True
    permissions = [None]


class RxLogin(RxMessage):
    type = 'login'
    proto = Login
    auth_required = False
    permissions = [None]


class RxAppleLogin(RxMessage):
    type = 'apple-login'
    proto = AppleLogin
    auth_required = False
    permissions = [None]


class RxVerifyToken(RxMessage):
    type = 'verify-token'
    proto = VerifyToken
    auth_required = True
    permissions = [None]


class TxTokenInvalid(TxMessage):
    type = 'token-invalid'
    proto = TokenInvalid
    auth_required = True
    permissions = [None]


class TxFormErrors(TxMessage):
    type = 'form-errors'
    proto = FormErrors
    auth_required = True
    permissions = [None]


class TxUploadStartSlot(TxMessage):
    type = 'upload-start'
    proto = UploadStartSlot
    auth_required = True
    permissions = [None]


class TxUploadProgress(TxMessage):
    type = 'upload-progress'
    proto = UploadProgress
    auth_required = True
    permissions = [None]


class TxUploadDone(TxMessage):
    type = 'upload-done'
    proto = UploadDone
    auth_required = True
    permissions = [None]


class RxUploadStart(RxMessage):
    type = 'upload-start'
    proto = UploadStart
    auth_required = True
    permissions = [None]


class RxDeleteFile(RxMessage):
    type = 'delete-file'
    proto = DeleteFile
    auth_required = True
    permissions = [None]


class RxUploadEnd(RxMessage):
    type = 'upload-end'
    proto = UploadEnd
    auth_required = True
    permissions = [None]


class TxAck(TxMessage):
    type = 'ack'
    proto = Ack
    auth_required = True
    permissions = [None]


class TxUpgradeApiVersion(TxMessage):
    type = 'upgrade-api-version'
    proto = UpgradeApiVersion
    auth_required = True
    permissions = [None]


class TxUploadTask(TxMessage):
    type = 'upload-task'
    proto = UploadTask
    auth_required = True
    permissions = [None]


class TxWasteContainers(TxMessage):
    type = 'waste-containers'
    proto = WasteContainers
    auth_required = True
    permissions = [None]


class RxScanWasteContainer(RxMessage):
    type = 'scan-container'
    proto = ScanWasteContainer
    auth_required = True
    permissions = [None]


class TxScannedWasteContainer(TxMessage):
    type = 'scanned-container'
    proto = ScannedWasteContainer
    auth_required = True
    permissions = [None]


class RxLoadContainerInfo(RxMessage):
    type = 'load-container-info'
    proto = LoadContainerInfo
    auth_required = True
    permissions = [None]


class TxContainerInfo(TxMessage):
    type = 'container-info'
    proto = ContainerInfo
    auth_required = True
    permissions = [None]


class RxLoadContainerMass(RxMessage):
    type = 'load-container-mass'
    proto = LoadContainerMass
    auth_required = True
    permissions = [None]


class TxContainerMass(TxMessage):
    type = 'container-mass'
    proto = ContainerMass
    auth_required = True
    permissions = [None]


class RxLoadContainerImpact(RxMessage):
    type = 'load-container-impact'
    proto = LoadContainerImpact
    auth_required = True
    permissions = [None]


class TxContainerImpact(TxMessage):
    type = 'container-impact'
    proto = ContainerImpact
    auth_required = True
    permissions = [None]


class RxLoadContainerPurchases(RxMessage):
    type = 'load-container-purchases'
    proto = LoadContainerPurchases
    auth_required = True
    permissions = [None]


class TxContainerPurchases(TxMessage):
    type = 'container-purchases'
    proto = ContainerPurchases
    auth_required = True
    permissions = [None]


class RxGetHomeInfo(RxMessage):
    type = 'get-home-info'
    proto = GetHomeInfo
    auth_required = True
    permissions = [None]


class TxHomeInfo(TxMessage):
    type = 'home-info'
    proto = HomeInfo
    auth_required = True
    permissions = [None]


class RxLoadRewards(RxMessage):
    type = 'load-rewards'
    proto = LoadRewards
    auth_required = True
    permissions = [None]


class TxRewards(TxMessage):
    type = 'rewards'
    proto = Rewards
    auth_required = True
    permissions = [None]


class RxGetMagicLink(RxMessage):
    type = 'get-magic-link'
    proto = GetMagicLink
    auth_required = True
    permissions = [None]


class RxLoginMagicLink(RxMessage):
    type = 'login-magic-link'
    proto = LoginMagicLink
    auth_required = True
    permissions = [None]


class TxLoginMagicLinkStatus(TxMessage):
    type = 'login-magic-link-status'
    proto = LoginMagicLinkStatus
    auth_required = True
    permissions = [None]


class TxUserProfileData(TxMessage):
    type = 'user-profile-data'
    proto = UserProfileData
    auth_required = True
    permissions = [None]
